File: coordinator/protocol/process_memory_telemetry.go
package protocol

// ProcessMemoryTelemetry is one coherent process admission observation. Charged
// C includes materialized M; U+(C-M) is the projection, never U+C. It is diagnostic
// only. Omitted instrumentation and OS availability remain unknown.
type ProcessMemoryTelemetry struct {
	Generation             uint64  `json:"generation"`
	SampleSeq              uint64  `json:"sample_seq"`
	SampleAgeMS            uint64  `json:"sample_age_ms"`
	PolicyEpoch            uint64  `json:"policy_epoch"`
	CapBytes               uint64  `json:"cap_bytes"`
	ActivationReserveBytes uint64  `json:"activation_reserve_bytes"`
	ActiveBytes            uint64  `json:"active_bytes"`
	CacheBytes             uint64  `json:"cache_bytes"`
	ChargedBytes           uint64  `json:"charged_bytes"`
	MaterializedBytes      uint64  `json:"materialized_bytes"`
	UnmaterializedBytes    uint64  `json:"unmaterialized_bytes"`
	RemainingBytes         uint64  `json:"remaining_bytes"`
	CommitmentDebtBytes    uint64  `json:"commitment_debt_bytes"`
	OwnerCount             uint64  `json:"owner_count"`
	ClosingOwnerCount      uint64  `json:"closing_owner_count"`
	SystemAvailableBytes   *uint64 `json:"system_available_bytes,omitempty"`
}

func (s *ProcessMemoryTelemetry) Clone() *ProcessMemoryTelemetry {
	if s == nil {
		return nil
	}
	result := *s
	result.SystemAvailableBytes = clonePtr(s.SystemAvailableBytes)
	return &result
}

File: provider-swift/Sources/ProviderCore/Protocol/ProcessMemoryTelemetry.swift
import Foundation

/// One coherent process ledger observation. C and M overlap: admission uses
/// active + cache + (charged - materialized), subject to cap/reserve/OS limits.
/// This reports ownership; it never grants capacity or authorizes cache reuse.
public struct ProcessMemoryTelemetry: Codable, Sendable, Equatable {
    public var generation: UInt64 = 0
    public var sampleSeq: UInt64 = 0
    public var sampleAgeMs: UInt64 = 0
    public var policyEpoch: UInt64 = 0
    public var capBytes: UInt64 = 0
    public var activationReserveBytes: UInt64 = 0
    public var activeBytes: UInt64 = 0
    public var cacheBytes: UInt64 = 0
    public var chargedBytes: UInt64 = 0
    public var materializedBytes: UInt64 = 0
    public var unmaterializedBytes: UInt64 = 0
    public var remainingBytes: UInt64 = 0
    public var commitmentDebtBytes: UInt64 = 0
    public var ownerCount: UInt64 = 0
    public var closingOwnerCount: UInt64 = 0
    public var systemAvailableBytes: UInt64?
    // Native clock is local-only. Heartbeat stamping ages the captured value
    // without resampling the ledger or treating another send as new evidence.
    var capturedUptimeNanoseconds: UInt64?

    enum CodingKeys: String, CodingKey {
        case generation = "generation"
        case sampleSeq = "sample_seq"
        case sampleAgeMs = "sample_age_ms"
        case policyEpoch = "policy_epoch"
        case capBytes = "cap_bytes"
        case activationReserveBytes = "activation_reserve_bytes"
        case activeBytes = "active_bytes"
        case cacheBytes = "cache_bytes"
        case chargedBytes = "charged_bytes"
        case materializedBytes = "materialized_bytes"
        case unmaterializedBytes = "unmaterialized_bytes"
        case remainingBytes = "remaining_bytes"
        case commitmentDebtBytes = "commitment_debt_bytes"
        case ownerCount = "owner_count"
        case closingOwnerCount = "closing_owner_count"
        case systemAvailableBytes = "system_available_bytes"
    }

    func agedForHeartbeat(now: UInt64) -> Self? {
        guard let capturedUptimeNanoseconds else { return self }
        guard now >= capturedUptimeNanoseconds else { return nil }
        var result = self
        result.sampleAgeMs = max(sampleAgeMs,
            min((now - capturedUptimeNanoseconds) / 1_000_000, (1 << 53) - 1))
        return result
    }
}
